types/mempool/priority_nonce.go

```diff
@@ -12,137 +12,169 @@ import (
 )
 
 var (
-	_ Mempool  = (*PriorityNonceMempool)(nil)
-	_ Iterator = (*PriorityNonceIterator)(nil)
+	_ Mempool  = (*PriorityNonceMempool[int64])(nil)
+	_ Iterator = (*PriorityNonceIterator[int64])(nil)
 )
 
-// PriorityNonceMempool is a mempool implementation that stores txs
-// in a partially ordered set by 2 dimensions: priority, and sender-nonce
-// (sequence number). Internally it uses one priority ordered skip list and one
-// skip list per sender ordered by sender-nonce (sequence number). When there
-// are multiple txs from the same sender, they are not always comparable by
-// priority to other sender txs and must be partially ordered by both sender-nonce
-// and priority.
-type PriorityNonceMempool struct {
-	priorityIndex  *skiplist.SkipList
-	priorityCounts map[int64]int
-	senderIndices  map[string]*skiplist.SkipList
-	scores         map[txMeta]txMeta
-	onRead         func(tx sdk.Tx)
-	txReplacement  func(op, np int64, oTx, nTx sdk.Tx) bool
-	maxTx          int
-}
-
-type PriorityNonceIterator struct {
-	senderCursors map[string]*skiplist.Element
-	nextPriority  int64
-	sender        string
-	priorityNode  *skiplist.Element
-	mempool       *PriorityNonceMempool
-}
-
-// txMeta stores transaction metadata used in indices
-type txMeta struct {
-	// nonce is the sender's sequence number
-	nonce uint64
-	// priority is the transaction's priority
-	priority int64
-	// sender is the transaction's sender
-	sender string
-	// weight is the transaction's weight, used as a tiebreaker for transactions with the same priority
-	weight int64
-	// senderElement is a pointer to the transaction's element in the sender index
-	senderElement *skiplist.Element
-}
-
-// txMetaLess is a comparator for txKeys that first compares priority, then weight,
-// then sender, then nonce, uniquely identifying a transaction.
-//
-// Note, txMetaLess is used as the comparator in the priority index.
-func txMetaLess(a, b any) int {
-	keyA := a.(txMeta)
-	keyB := b.(txMeta)
-	res := skiplist.Int64.Compare(keyA.priority, keyB.priority)
-	if res != 0 {
-		return res
+type (
```

Comment:
The `type (` block like this is not commonly used in Go. It adds indentation and impedes `grep type TYPENAME`. Please keep the individual type declarations.

Reply:
When there are multiple types in a file, I use `()` to group them to make it easier for the reader.